channeldb/mp_payment.go

```diff
@@ -0,0 +1,137 @@
+	AttemptTime time.Time
+
+	// Settle is the preimage of a successful payment. This serves as a
+	// proof of payment. It will only be non-nil for settled payments.
+	//
+	// NOTE: Can be nil if payment is not settled.
+	Settle *HTLCSettleInfo
+
+	// Fail is a failure reason code indicating the reason the payment
+	// failed. It is only non-nil for failed payments.
+	//
+	// NOTE: Can be nil if payment is not failed.
+	Failure *HTLCFailInfo
+}
+
+// HTLCSettleInfo encapsulates the information that augments an HTLCAttempt int
+// he event that the HTLC is successful.
+type HTLCSettleInfo struct {
+	// Preimage is the preimage of a successful HTLC. This serves as a proof
+	// of payment.
+	Preimage lntypes.Preimage
+
+	// SettleTime is the time at which this HTLC was settled.
+	SettleTime time.Time
+}
+
+// HTLCFailInfo encapsulates the information that augments an HTLCAttempt in the
+// event that the HTLC fails.
+type HTLCFailInfo struct {
+	// Reason is the failure reason code indicating the reason the HTLC
+	// failed.
+	Reason FailureReason
+
+	// FailTime is the time at which this HTLC was failed.
+	FailTime time.Time
+}
+
+// MPPayment is a wrapper around a payment's MPPaymentCreationInfo and
+// HTLCAttempts. All payments will have the MPPPaymentCreationInfo set, any
+// HTLCs made in attempts to be completed will populated in the HTLCs slice.
+// Each populated HTLCAttempt represents an attempted HTLC, each of which may
+// have the associated Settle or Fail struct populated if the HTLC is no longer
+// in-flight.
+type MPPayment struct {
+	// sequenceNum is a unique identifier used to sort the payments in
+	// order of creation.
+	sequenceNum uint64
+
+	// Status is the current PaymentStatus of this payment.
+	Status PaymentStatus
+
+	// Info holds all static information about this payment, and is
+	// populated when the payment is initiated.
+	Info *MPPaymentCreationInfo
+
+	// HTLCs holds the information about individual HTLCs that we send in
+	// order to make the payment.
+	HTLCs []HTLCAttempt
+}
+
+// Route returns the route used to complete the payment, or the final route
+// attempted if the payment failed or is in-flight.
+func (p *MPPayment) Route() route.Route {
+	var someRoute route.Route
+	for _, htlc := range p.HTLCs {
+		// Return immediately if we find a successful route.
+		if htlc.Settle != nil {
+			return htlc.Route
+		}
+
+		// Otherwise stash this route which will be returned if their
+		// are no successful htlcs.
+		someRoute = htlc.Route
+	}
+	return someRoute
+}
+
+// PaymentPreimage returns the preimage learned as a result of the payment if it
+// was successful, or the zero-preimage if the payment failed or is still
+// in-flight.
+func (p *MPPayment) PaymentPreimage() lntypes.Preimage {
+	for _, htlc := range p.HTLCs {
+		// Return the preimage if we find a successful HTLC.
+		if htlc.Settle != nil {
+			return htlc.Settle.Preimage
+		}
+	}
+
+	return lntypes.Preimage{}
+}
```

Comment:
For review, I'd like it more to also have code using this struct in this commit, prevent flipping back and forth between two commits. Personal pref. though